utils/message-generator/src/main.rs

```diff
@@ -17,87 +17,254 @@ use roles_logic_sv2::{common_messages_sv2::SetupConnectionSuccess, parsers::AnyM
+/// present if the `"role": "downstream"` or the `"role": "proxy"` key-value pair is present.
 #[derive(Debug, Clone)]
 struct Downstream {
+    /// Host endpoint address.
     addr: SocketAddr,
+    /// Host endpoint pubkey. If present, a noise connection will be used, otherwise a plain
+    /// connection will be used.
     key: Option<EncodedEd25519PublicKey>,
 }
 
+/// Represents the message identifiers of the messages to execute, the expected responses of each
+/// message, and the endpoint information of the role being mocked.
+/// Serialized from the `"actions"` key-value pair in the `test.json` configuration file.
 #[derive(Debug)]
 pub struct Action<'a> {
+    /// `PoolMessages` messages to execute serialized into a data frame.
     messages: Vec<EitherFrame<AnyMessage<'a>>>,
+    /// Expected message response results of each `PoolMessages` in `messages`.
     result: Vec<ActionResult>,
+    /// Role being mocked. Can be a downstream client role, a proxy role, or an upstream server
+    /// role.
     role: Role,
 }
 
+/// Represents a shell execution command defined in either the `"setup_commmands"`,
+/// `"execution_commands"` or `"cleanup_commmands"` key-value pair is present in the `test.json`
+/// configuration file.
 #[derive(Debug, Serialize, Deserialize)]
 pub struct Command {
+    /// Binary to call (first argument of shell command).
     command: String,
+    /// Flags or commands in call (remaining arguments in shell command).
     args: Vec<String>,
+    /// TODO: ??
     conditions: ExternalCommandConditions,
 }
 
+/// Represents all of the parsed contents from the `test.json` configuration file, ready for
+/// execution.
 #[derive(Debug)]
 pub struct Test<'a> {
+    /// Represents the result of a specified `Action` defined in `test.json` configuration file as
+    /// the `"results"` key-pair within the `"actions"` key value. It represents the expected
+    /// response(s) to a specified `PoolMessages` message.
     actions: Vec<Action<'a>>,
+    /// Represents the endpoint connection information of an upstream server role being mocked,
+    /// defined in the `"upstream"` key-value pair in the `test.json` configuration file. Present if the
+    /// `"role": "upstream"` or the `"role": "proxy"` key-value pair is present.
     as_upstream: Option<Upstream>,
+    /// Represents the endpoint connection information of an downstream server role being mocked,
+    /// defined in the `"downstream"` key-value pair in the `test.json` configuration file. Present if
+    /// the `"role": "downstream"` or the `"role": "proxy"` key-value pair is present.
     as_dowstream: Option<Downstream>,
+    /// Represents the `"setup_commands"` key-value pair in the `test.json` configuration file that
+    /// contains shell commands to run before any tests are executed.
     setup_commmands: Vec<Command>,
+    /// Represents the `"execution_commands"` key-value pair in the `test.json` configuration file
+    /// that TODO: ???.
```

Comment:
@Fi3 what purpose does the `"execution_commands"` key have? Is this to run a shell command mid-message passing during a test? Can you give an example of when one would want to do this, if so?

Reply:
This is for shell commands that need to be run after a connection has been opened between two roles. For example, if using the MG to test a proxy, starting up a Downstream mining device could be done here--after the proxy has had a chance to connect to its Upstream.